### PR TITLE
Loader simplification

### PR DESCRIPTION
**Pull request by jaheba**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

File: src/gluonts/dataset/loader.py [hunk 1]
```diff
@@ -30,7 +28,34 @@
...
+
+    batches = batcher(dataset, batch_size)
+    batches = map(stack_fn, batches)
+    raw_batches = map(_encode, batches)
+
+    for raw in raw_batches:
+        try:
+            batch_queue.put(raw)
+        except (EOFError, BrokenPipeError):
+            return
```
Comment (lostella) [inline src/gluonts/dataset/loader.py:55]: Neat!

**Review by lostella — COMMENTED** [review]
This looks much nicer, thanks. It looks like some type annotation is making mypy unhappy, maybe it's the one you added to TrainDataLoader?

File: src/gluonts/dataset/loader.py [hunk 2]
```diff
@@ -12,226 +12,141 @@
...
+
+    def __call__(self, data, is_train):
+        yield from batcher(data, self.batch_size)
 
 
 def TrainDataLoader(
     dataset: Dataset,
     *,
-    transform: Transformation,
+    transform: Transformation = Identity(),
```
Comment (lostella) [inline src/gluonts/dataset/loader.py:144]: I like this a lot

File: src/gluonts/dataset/loader.py [hunk 3]
```diff
@@ -282,33 +197,34 @@ def TrainDataLoader(
...
-    data_iterable = (
-        PseudoShuffled(
-            transformed_dataset, shuffle_buffer_length=shuffle_buffer_length
+    dataset = Cyclic(dataset)
+
+    if shuffle_buffer_length:
+        dataset = PseudoShuffled(dataset, shuffle_buffer_length)
+
+    transform += Batch(batch_size=batch_size) + AdhocTransform(stack_fn)
+    batches: DataLoader = transform.apply(dataset, is_train=True)
```
Comment (lostella) [inline src/gluonts/dataset/loader.py:206]: This is fine, but isn't the "native" python alternative better?

```python
batches = batcher(transform.apply(dataset, is_train=True), batch_size)
batches: DataLoader = map(stack_fn, batches)
```

**Review by lostella — APPROVED** [review]
A net loss of ~90 lines is always good to merge
